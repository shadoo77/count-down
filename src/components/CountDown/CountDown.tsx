import { useCountDown } from '../../hooks/useCountDown';

function ExpiredDate() {
  return (
    <div className="expired-notice">
      <span>Expired!!!</span>
      <p>Please select a future date and time.</p>
    </div>
  );
}

function CountDown() {
  const { timeleft, isExpired } = useCountDown();
  const { days, hours, minutes, seconds } = timeleft;

  if (isExpired) {
    return <ExpiredDate />;
  }

  return (
    <section
      className='count-down'
      data-testid="count-down"
    >
      <div>
        <span>{days}</span>
        <span><small>days</small></span>
      </div>
      <div>
        <span>{hours}</span>
        <span><small>hours</small></span>
      </div>
      <div>
        <span>{minutes}</span>
        <span><small>minutes</small></span>
      </div>
      <div>
        <span>{seconds}</span>
        <span><small>seconds</small></span>
      </div>
    </section>
  );
}

export default CountDown;
