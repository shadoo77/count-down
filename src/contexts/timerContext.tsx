import React, { createContext, useState, useCallback } from 'react';

export interface TimeLeft {
  id: number;
  name: string;
}

export interface IState {
  date: string;
  startCountDown: boolean;
}

interface ITimerContext extends IState {
  stateUpdater: React.Dispatch<Partial<IState>>;
  resetState: React.DispatchWithoutAction;
}

const initialState: IState = {
  date: '',
  startCountDown: false,
};

const TimerContext = createContext<ITimerContext | undefined>(
  undefined
);

interface IProviderProps {
  children: React.ReactElement;
}

export function TimerProvider({ children }: IProviderProps) {
  const [state, setState] = useState(initialState);

  const stateUpdater = useCallback((obj: Partial<IState>) => {
    setState((prevState: IState) => ({
      ...prevState,
      ...obj
    }));
  }, []);

  const resetState = useCallback(() => {
    setState(initialState);
  }, []);

  return (
    <TimerContext.Provider
      value={{
        ...state,
        stateUpdater,
        resetState
      }}
    >
      {children}
    </TimerContext.Provider>
  );
}

export function useTimer() {
  const context = React.useContext(TimerContext);
  if (context === undefined) {
    throw new Error('useTimer must be used within a TimerProvider');
  }
  return context;
}
