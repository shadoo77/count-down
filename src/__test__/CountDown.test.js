import renderer from 'react-test-renderer';
import { render, screen, cleanup } from '@testing-library/react';
import CountDown from '../components/CountDown';
import { TimerProvider } from '../contexts/timerContext';

const CountDownElement = (
  <TimerProvider>
    <CountDown />
  </TimerProvider>
);

afterEach(() => {
  cleanup();
});

it('CountDown is rendering successfully', () => {
  render(CountDownElement);
  const countdownElement = screen.getByTestId('count-down');
  expect(countdownElement).toBeInTheDocument();
  expect(countdownElement).toHaveClass('count-down');
  expect(countdownElement).toContainHTML('div');
  expect(countdownElement.childElementCount).toEqual(4);
});

it('matches snapshot', () => {
  const component = renderer.create(CountDownElement);
  const tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});
