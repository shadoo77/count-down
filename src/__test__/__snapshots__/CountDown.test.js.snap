// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<section
  className="count-down"
  data-testid="count-down"
>
  <div>
    <span>
      00
    </span>
    <span>
      <small>
        days
      </small>
    </span>
  </div>
  <div>
    <span>
      00
    </span>
    <span>
      <small>
        hours
      </small>
    </span>
  </div>
  <div>
    <span>
      00
    </span>
    <span>
      <small>
        minutes
      </small>
    </span>
  </div>
  <div>
    <span>
      00
    </span>
    <span>
      <small>
        seconds
      </small>
    </span>
  </div>
</section>
`;
